// @aws-amplify/graphql-transformer-core/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransformerContractError = exports.MappingTemplate = exports.getAppSyncServiceExtraDirectives = exports.TransformerPluginBase = exports.TransformerModelEnhancerBase = exports.TransformerModelBase = exports.DirectiveWrapper = exports.collectDirectivesByTypeNames = exports.collectDirectives = exports.ConflictHandlerType = exports.GraphQLTransform = void 0;
const graphql_1 = require("graphql");
const validation_1 = require("./transformation/validation");
var transformation_1 = require("./transformation");
Object.defineProperty(exports, "GraphQLTransform", { enumerable: true, get: function () { return transformation_1.GraphQLTransform; } });
var index_1 = require("./config/index");
Object.defineProperty(exports, "ConflictHandlerType", { enumerable: true, get: function () { return index_1.ConflictHandlerType; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "collectDirectives", { enumerable: true, get: function () { return utils_1.collectDirectives; } });
Object.defineProperty(exports, "collectDirectivesByTypeNames", { enumerable: true, get: function () { return utils_1.collectDirectivesByTypeNames; } });
Object.defineProperty(exports, "DirectiveWrapper", { enumerable: true, get: function () { return utils_1.DirectiveWrapper; } });
var transformer_plugin_base_1 = require("./transformation/transformer-plugin-base");
Object.defineProperty(exports, "TransformerModelBase", { enumerable: true, get: function () { return transformer_plugin_base_1.TransformerModelBase; } });
Object.defineProperty(exports, "TransformerModelEnhancerBase", { enumerable: true, get: function () { return transformer_plugin_base_1.TransformerModelEnhancerBase; } });
Object.defineProperty(exports, "TransformerPluginBase", { enumerable: true, get: function () { return transformer_plugin_base_1.TransformerPluginBase; } });
const getAppSyncServiceExtraDirectives = () => {
    return graphql_1.print(validation_1.EXTRA_DIRECTIVES_DOCUMENT);
};
exports.getAppSyncServiceExtraDirectives = getAppSyncServiceExtraDirectives;
var cdk_compat_1 = require("./cdk-compat");
Object.defineProperty(exports, "MappingTemplate", { enumerable: true, get: function () { return cdk_compat_1.MappingTemplate; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "TransformerContractError", { enumerable: true, get: function () { return errors_1.TransformerContractError; } });
